Content.Server/Body/Systems/InternalsSystem.cs

```diff
@@ -1,24 +1,175 @@
     {
-        if (component.AreInternalsWorking())
+        if (AreInternalsWorking(component))
         {
             var gasTank = Comp<GasTankComponent>(component.GasTankEntity!.Value);
-            args.Gas = gasTank.RemoveAirVolume(Atmospherics.BreathVolume);
+            args.Gas = _gasTank.RemoveAirVolume(gasTank, Atmospherics.BreathVolume);
+            // TODO: Should listen to gas tank updates instead I guess?
+            _alerts.ShowAlert(component.Owner, AlertType.Internals, GetSeverity(component));
+        }
+    }
+    public void DisconnectBreathTool(InternalsComponent component)
+    {
+        var old = component.BreathToolEntity;
+        component.BreathToolEntity = null;
+
+        if (TryComp(old, out BreathToolComponent? breathTool) )
+        {
+            _atmos.DisconnectInternals(breathTool);
+            DisconnectTank(component);
+        }
+
+        _alerts.ShowAlert(component.Owner, AlertType.Internals, GetSeverity(component));
+    }
+
+    public void ConnectBreathTool(InternalsComponent component, EntityUid toolEntity)
+    {
+        if (TryComp(component.BreathToolEntity, out BreathToolComponent? tool))
+        {
+            _atmos.DisconnectInternals(tool);
+        }
+
+        component.BreathToolEntity = toolEntity;
+        _alerts.ShowAlert(component.Owner, AlertType.Internals, GetSeverity(component));
+    }
+
+    public void DisconnectTank(InternalsComponent? component)
+    {
+        if (component == null) return;
+
+        if (TryComp(component.GasTankEntity, out GasTankComponent? tank))
+        {
+            _gasTank.DisconnectFromInternals(tank, component.Owner);
+        }
+
+        component.GasTankEntity = null;
+        _alerts.ShowAlert(component.Owner, AlertType.Internals, GetSeverity(component));
+    }
+
+    public bool TryConnectTank(InternalsComponent component, EntityUid tankEntity)
+    {
+        if (component.BreathToolEntity == null)
+            return false;
+
+        if (TryComp(component.GasTankEntity, out GasTankComponent? tank))
+        {
+            _gasTank.DisconnectFromInternals(tank, component.Owner);
+        }
+
+        component.GasTankEntity = tankEntity;
+        _alerts.ShowAlert(component.Owner, AlertType.Internals, GetSeverity(component));
+        return true;
+    }
+
+    public bool AreInternalsWorking(InternalsComponent component)
+    {
+        return TryComp(component.BreathToolEntity, out BreathToolComponent? breathTool) &&
+               breathTool.IsFunctional &&
+               TryComp(component.GasTankEntity, out GasTankComponent? gasTank) &&
+               gasTank.Air != null;
+    }
+
+    private short GetSeverity(InternalsComponent component)
+    {
+        if (component.BreathToolEntity == null || !AreInternalsWorking(component)) return 2;
+
+        if (TryComp<GasTankComponent>(component.GasTankEntity, out var gasTank) && gasTank.Air.Volume < Atmospherics.BreathVolume)
+            return 0;
+
+        return 1;
+    }
```

Comment:
Not really sure what to do on checking volume / moles as I couldn't tell if there was an easy way to check if the air was enough?

Reply:
I guess this should be checking the release pressure, and making sure that there is enough gas to fill a breath-volume with that much pressure? So something like `tankPressure * breathVolume < tankMoles * tankTemperature`.

Reply:
Didn't work for me.

Reply:
Ah, I forgot about `Atmospherics.R`, So it should've been.
```
if (gasTank.OutputPressure * Atmospherics.BreathVolume > gasTank.Air.TotalMoles * Atmospherics.R * gasTank.Air.Temperature)
        return 0;
```

But even that doesn't work, because the output pressure is constantly getting lowered as the tank runs out of air, so the condition is still always true as `tankVolume > BreathVolume`.

When is the alert status actually meant to change / what is it supposed to indicate? Maye the configured release pressure just shouldn't auto-lower as the pressure goes down, and the alert should just show up when there available pressure is less than the requested amount?

Reply:
I don't know how exactly it works in ss13 but more or less when the person starts gasping and the tank has insufficient air it changes the alert.

Reply:
Unless you want to get the warning indicator working first, it should be fine to just merge the PR as is.

I dunno the details about how it works in ss13, but how I imagine the indicator/status should work in an ideal world is that it should tell you you are going to start suffocating 10-30 seconds before that actually happens and you start taking damage. But implementing that is a PITA because that depends on the gas mix, pressure, and species metabolism. A less-than-requested pressure warning isn't as useful, but probably much easier.